File: prediction_models/tile_concat_wy/input/spine.py
import cv2
import numpy as np
import collections
from collections import OrderedDict
from skimage.measure import label
from skimage.morphology import reconstruction, thin, skeletonize
from PIL import Image, ImageDraw


def skeleton_endpoints(skel):
    h,w = skel.shape
    # Find row and column locations that are non-zero
    (rows, cols) = np.nonzero(skel)
    # Initialize empty list of co-ordinates
    skel_coords = []
    # For each non-zero pixel...
    for (r, c) in zip(rows, cols):
        # Extract an 8-connected neighbourhood
        (col_neigh, row_neigh) = np.meshgrid(np.array([c - 1, c, c + 1]), np.array([r - 1, r, r + 1]))
        # Cast to int to index into image
        col_neigh = col_neigh.astype('int')
        row_neigh = row_neigh.astype('int')
        # Convert into a single 1D array and check for non-zero locations
        pix_neighbourhood = skel[row_neigh.clip(0,h-1), col_neigh.clip(0,w-1)].ravel() != 0
        # If the number of non-zero locations equals 2, add this to
        # our list of co-ordinates
        if np.sum(pix_neighbourhood) == 2:
            skel_coords.append((r, c))
    return skel_coords

def bfs(grid, start, goal):
    width, height = grid.shape
    queue = collections.deque([[start]])
    seen = set([start])
    steps = np.zeros_like(grid)
    while queue:
        for i in range(len(queue)):
            path = queue.popleft()
            x, y = path[-1]
            if (x, y) == goal:
                return steps, path
            for x2, y2 in (
            (x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1), (x + 1, y + 1), (x - 1, y - 1), (x + 1, y - 1),
            (x - 1, y + 1)):
                if 0 <= x2 < width and 0 <= y2 < height and grid[x2][y2] == 1 and (x2, y2) not in seen:
                    queue.append(path + [(x2, y2)])
                    seen.add((x2, y2))
                    steps[x2, y2] = steps[x, y] + 1
    return -1, None

def spine(img, **kwargs):
    result = OrderedDict()
    thresh = kwargs['thresh'] if 'thresh' in kwargs else 0.9
    bthresh = kwargs['bthresh'] if 'bthresh' in kwargs else 20
    min_size = kwargs['min_size'] if 'min_size' in kwargs else 40
    tan_size = kwargs['tan_size'] if 'tan_size' in kwargs else 4
    patch_size = kwargs['patch_size'] if 'patch_size' in kwargs else 128
    step_size = kwargs['step_size'] if 'step_size' in kwargs else int(0.8 * patch_size)
    slide_thresh = kwargs['slide_thresh'] if 'slide_thresh' in kwargs else 0.6
    overlap_thresh = kwargs['overlap_thresh'] if 'overlap_thresh' in kwargs else 0.2
    h_step_size = kwargs['h_step_size'] if 'h_step_size' in kwargs else 0.5

    ## first step: find the spine
    im2 = img
    im3 = np.mean(im2, 2)  # returns the mean over channel dimension
    im4 = im3 < thresh * np.max(im3)  # True: content, False white pixel
    im4[im2[:, :, 1] > im3] = 0  # areas that are too greenish
    im4[im2[:, :, 2] > im2[:, :, 0] + bthresh] = 0  # areas that are too blueish
    im4 = im4.astype('int')

    im5 = im4

    seed = np.copy(im5)
    seed[1:-1, 1:-1] = np.max(im5)
    im6 = reconstruction(seed, im5, method='erosion')  # image erosion -> the best mask
    mask = im6.astype('int')

    im7 = skeletonize(im6)  # find skelenton
    im7 = thin(im7)  # shrink skelenton

    spine = np.zeros_like(im7).astype('float')
    cc, num_blob = label(mask, connectivity=2, return_num=True)
    stespath = []
    location = []
    IOU = []
    polymasks = np.zeros_like(im7).astype('int')
    rad = patch_size / 2.0
    rad3 = patch_size * 2
    for prop in range(1, num_blob + 1):
        #         print("{}/{}".format(prop,num_blob))
        tim = np.zeros_like(im7).astype(np.uint8)
        tim[cc == prop] = 1
        if np.sum(tim) < min_size:
            continue
        skel = skeletonize(tim)
        skel = thin(skel)
        E = skeleton_endpoints(skel)
        maxval = -1
        for i in range(min(5, len(E))):
            for j in range(min(5, len(E))):
                tim2, path = bfs(skel.astype(np.int64), list(E)[i], list(E)[-j])
                if np.max(tim2) > maxval:
                    maxval = np.max(tim2)
                    stespath = path
        tim1 = np.zeros_like(im7).astype('float')
        for i in range(len(stespath)):
            tim1[stespath[i]] = 1
        spine = spine + tim1  # spine -> spine image

        ## second step: find the tile location:
        E = skeleton_endpoints(tim1)
        if len(E) < 2:
            continue
        _, pts = bfs(tim, E[0], E[1])  # find the shortest path on the spine
        pvt = rad
        if not pts:
            continue
        while pvt < len(pts) - tan_size:
            #             print(pvt)
            ## for each pvt 1: find perpendicular line and its width
            p1 = pts[int(pvt - tan_size)]
            p2 = pts[int(pvt + tan_size)]
            tan = (np.array(p2) - np.array(p1)) / np.sqrt(np.sum((np.array(p2) - np.array(p1)) ** 2))
            x2 = pts[int(pvt)][0] + [-rad3 * tan[1], rad3 * tan[1]];
            y2 = pts[int(pvt)][1] + [rad3 * tan[0], -rad3 * tan[0]];
            x2 = np.linspace(x2[0], x2[1])
            y2 = np.linspace(y2[0], y2[1])
            y2 = y2[np.multiply(x2 >= 0, x2 < tim.shape[0])]
            x2 = x2[np.multiply(x2 >= 0, x2 < tim.shape[0])]
            x2 = x2[np.multiply(y2 >= 0, y2 < tim.shape[1])]
            y2 = y2[np.multiply(y2 >= 0, y2 < tim.shape[1])]  # x2, y2 represents the line perpendicular to the spline
            valid = mask[x2.round().astype('int').clip(0, tim.shape[0] - 1), y2.round().astype('int').clip(0, tim.shape[
                1] - 1)] == 1
            x2 = x2[valid]
            y2 = y2[valid]
            if np.sum(valid) < 2:
                pvt = int(pvt + step_size)
                continue
            x2 = [x2[0], x2[-1]]
            y2 = [y2[0], y2[-1]]  # x2, y2 is now the boundary points of the perpendicular line
            dist = np.sqrt((x2[1] - x2[0]) ** 2 + (y2[1] - y2[0]) ** 2)  # tissue width of current pvt point

            ## for each width assign bx n based on patch width
            n = max(1, int(dist / (h_step_size * patch_size)))
            #             x3 = np.linspace(x2[0] + rad * tan[1], x2[1]-rad * tan[1], n)
            #             y3 = np.linspace(y2[0] + rad * tan[0], y2[1]-rad * tan[0], n)
            x3 = np.linspace(x2[0], x2[1], n)
            y3 = np.linspace(y2[0], y2[1], n)

            ## find the four corordinates of the patch and return the four cordinates
            mid = int(n / 2)
            for k in range(n):
                mod = int((k + 1) / 2)
                if k % 2 == 1:
                    pos = min(mid + mod, n - 1)
                else:
                    pos = max(mid - mod, 0)
                x = x3[pos] + [rad * tan[1], -rad * tan[1], -rad * tan[1] + 2 * rad * tan[0],
                               rad * tan[1] + 2 * rad * tan[0]]
                y = y3[pos] + [rad * tan[0], -rad * tan[0], -rad * tan[0] - 2 * rad * tan[1],
                               rad * tan[0] - 2 * rad * tan[1]]
                polymask = Image.new('L', (im7.shape[1], im7.shape[0]), 0)
                ImageDraw.Draw(polymask).polygon([y[0], x[0], y[1], x[1], y[2], x[2], y[3], x[3]], outline=1, fill=1)
                polymask = np.array(polymask)
                valid = np.multiply(mask, polymask)
                if np.sum(valid) > slide_thresh * np.sum(polymask):  # more than thresh are tissues
                    valid1 = np.multiply(polymask.astype('bool'), polymasks.astype('bool')).astype('bool')
                    if np.sum(valid1) < overlap_thresh * np.sum(
                            polymask):  # overlap with current selection smaller than thresh
                        # step_size = 0.8 * step_size
                        polymasks += polymask
                        location.append([[y[0], x[0]], [y[1], x[1]], [y[2], x[2]], [y[3], x[3]]])
                        IOU.append(np.sum(valid) / np.sum(polymask))
                    else:
                        # step_size = 1.1 * step_size
                        pass
            pvt = int(pvt + step_size)
    result['tile_location'] = location
    result['patch_mask'] = polymasks
    result['mask'] = mask
    result['spine'] = spine
    result['IOU'] = IOU
    return result

def remove_pen_marks(img, scale=1):
    # Define elliptic kernel
    kernel5x5 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (5, 5))
    # use cv2.inRange to mask pen marks (hardcoded for now)
    lower = np.array([0, 0, 0])
    upper = np.array([200, 255, 255])
    img_mask1 = cv2.inRange(img, lower, upper)

    # Use erosion and findContours to remove masked tissue (side effect of above)
    img_mask1 = cv2.erode(img_mask1, kernel5x5, iterations=4)
    img_mask2 = np.zeros(img_mask1.shape, dtype=np.uint8)
    _, contours, _ = cv2.findContours(img_mask1, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_SIMPLE)
    for contour in contours:
        #         print(contour.shape)
        x, y = contour[:, 0, 0], contour[:, 0, 1]
        w, h = x.max() - x.min(), y.max() - y.min()
        if w > 100 / scale and h > 100 / scale:
            cv2.drawContours(img_mask2, [contour], 0, 1, -1)
    # expand the area of the pen marks
    img_mask2 = cv2.dilate(img_mask2, kernel5x5, iterations=3)
    img_mask2 = (1 - img_mask2)

    # Mask out pen marks from original image
    img = cv2.bitwise_and(img, img, mask=img_mask2)

    img[img == 0] = 255
    return img, img_mask1, img_mask2

def tile(img, mask, location, iou, sz=256, N=36, scale = 8):
    result = []
    idxsort = np.argsort(iou)[::-1]
    for i in range(min(N, len(idxsort))):
        idx = idxsort[i]
        cnt = np.expand_dims(location[idx], 1) * scale
        cnt = cnt.astype('int')
        rect = cv2.minAreaRect(cnt)
        box = cv2.boxPoints(rect)
        box = np.int0(box)
        width = int(rect[1][0])
        height = int(rect[1][1])
        src_pts = box.astype("float32")
        dst_pts = np.array([[0, height - 1],
                            [0, 0],
                            [width - 1, 0],
                            [width - 1, height - 1]], dtype="float32")

        # the perspective transformation matrix
        M = cv2.getPerspectiveTransform(src_pts, dst_pts)

        # directly warp the rotated rectangle to get the straightened rectangle
        warped_img = cv2.warpPerspective(img, M, (width, height), borderValue = 255)
        warped_mask = cv2.warpPerspective(mask, M, (width, height), borderValue = 0)
        shape = warped_img.shape

        if shape[0] < sz or shape[1] < sz:
            pad0, pad1 = (sz - shape[0] % sz) % sz, (sz - shape[1] % sz) % sz
            warped_mask = np.pad(warped_mask, [[pad0 // 2, pad0 - pad0 // 2], [pad1 // 2, pad1 - pad1 // 2], [0, 0]], mode='constant', constant_values=0)
            warped_img = np.pad(warped_img, [[pad0 // 2, pad0 - pad0 // 2], [pad1 // 2, pad1 - pad1 // 2], [0, 0]], mode='constant', constant_values=255)

        result.append({'img': warped_img[:sz,:sz,:], 'mask': warped_mask[:sz,:sz,:], 'location': cnt})
    if len(idxsort) < N:
        for i in range(N - len(idxsort)):
            result.append({'img': 255 * np.ones((sz,sz,3)).astype(np.uint8),
                           'mask': np.zeros((sz,sz,3)).astype(np.uint8), 'location': None})
    return result

def tile_img(img, location, iou, sz=256, N=36, scale = 8):
    result = []
    idxsort = np.argsort(iou)[::-1]
    for i in range(min(N, len(idxsort))):
        idx = idxsort[i]
        cnt = np.expand_dims(location[idx], 1) * scale
        cnt = cnt.astype('int')
        rect = cv2.minAreaRect(cnt)
        box = cv2.boxPoints(rect)
        box = np.int0(box)
        width = int(rect[1][0])
        height = int(rect[1][1])
        src_pts = box.astype("float32")
        dst_pts = np.array([[0, height - 1],
                            [0, 0],
                            [width - 1, 0],
                            [width - 1, height - 1]], dtype="float32")

        # the perspective transformation matrix
        M = cv2.getPerspectiveTransform(src_pts, dst_pts)

        # directly warp the rotated rectangle to get the straightened rectangle
        warped_img = cv2.warpPerspective(img, M, (width, height))
        shape = warped_img.shape

        if shape[0] < sz or shape[1] < sz:
            pad0, pad1 = (sz - shape[0] % sz) % sz, (sz - shape[1] % sz) % sz
            warped_img = np.pad(warped_img, [[pad0 // 2, pad0 - pad0 // 2], [pad1 // 2, pad1 - pad1 // 2], [0, 0]], mode='constant', constant_values=255)

        result.append({'img': warped_img[:sz,:sz,:], 'location': cnt})
    if len(idxsort) < N:
        for i in range(N - len(idxsort)):
            result.append({'img': 255 * np.ones((sz,sz,3)).astype(np.uint8),
                           'location': None})
    return result



if __name__ == "__main__":
    img = None
    kwargs = {'thresh': 5}
    output = spine(img, **kwargs)
    print(output)

File: prediction_models/tile_concat_wy/input/spine.m
function [mask,counter2,locations]=slide_analysis2(im,scale,patch_size,step,x_ind1,x_ind2)
%% input images, scale, patch_size, step
if(nargin<2)
    scale=1;
end
if(nargin<3)
    patch_size=512;
end
if(nargin<4)
    step=patch_size;
end
if(nargin<5) %% what is x_ind1 and x_ind2
    x_ind1=1;
end
if(nargin<6)
    x_ind2=size(im,2);
end

colors={'yellow','red','green','blue','black'};

patch_width=patch_size/scale;
box_width=patch_width;
patch_height=patch_size/scale;
box_height=patch_height;
step_size=step/scale;
step_size2=step/scale;


rad=5; %strel radius in pixels
thresh=.9; %threshold for mask excluding the white space
min_size=40; %% smallest componets size
min_size2=min_size*scale;
tan_size=4;
bthresh=20;
hsvthresh=.1; % water stains and bubbles are removed by setting a lower bound on saturation which is 0.1
slide_thresh=.6; % patches that intersect with less than 60% intersection with the mask are removed
im2=im;
im3=mean(im2,3); %% returns the mean over third dimension
im4=im3<thresh*max(im3(:));
im4(im2(:,:,2)>im3)=0;
im4(im2(:,:,3)>im2(:,:,1)+bthresh)=0; %% what is this about
im2hsv=rgb2hsv(im2);
cc=bwconncomp(im4); % bwconncomp: Find connected components in binary image
im5=im4;
for i=1:length(cc.PixelIdxList)
    if length(cc.PixelIdxList{i})<min_size
        im5(cc.PixelIdxList{i})=0;
    end
end

### Not implementing
S=createSphericalStrel(rad); % morphological smoothing object
im6=convn(im5,S,'same')>0; % convolution im5 with S
im6(:,[1,end])=0; %% why is these two lines
im6([1,end],:)=0;
im7=1-(convn(1-im6,S,'same')>0);

im8=imfill(im7); % fill image regions and wholes
im8=bwmorph(im8,'skel',Inf); % morphological operations on binary images, find skelonton
im8 = bwmorph(im8,'thin',Inf); % prunning

cc=bwconncomp(im8);  % bwconncomp: Find connected components in binary image
im9=zeros(size(im8)); % or zeros with the same size
for i=1:length(cc.PixelIdxList)
    tim=zeros(size(im8));
    tim(cc.PixelIdxList{i})=1;
    E=find(bwmorph(tim,'endpoints')); % Find indices and values of nonzero elements
    maxind1=-1;
    maxind2=-1;
    maxval=-1;
    for j=1:length(E) %% find the longest endpoints in the image
        tim2=bwdistgeodesic(im8,E(j)); % compute the geodesic distance with each endpoints pair
        tim2(isinf(tim2))=0;
        if(max(tim2(:))>maxval)
            maxind1=E(j);
            maxval=max(tim2(:));
            maxind2=find(tim2==maxval);
        end
    end
    maxind2=maxind2(1);

    E=setdiff(E,[maxind1,maxind2]);
    while numel(E)>0 ## recursively set the longest skelonton
        tim(E)=0;
        E = find(bwmorph(tim, 'endpoints'));
        E=setdiff(E,[maxind1,maxind2]);
    end
    im9=im9+tim;
end
im9 = bwmorph(im9,'thin',Inf);
im9(:,1:x_ind1)=0;
im9(:,x_ind2:end)=0;

mask=im5;
mask(:,1:x_ind1)=0;
mask(:,x_ind2:end)=0;

imshow(im2,[]);
hold on;

cc=bwconncomp(im9);
counter2=0;
patches=[];
locations=[];
masks=[];
for i=1:length(cc.PixelIdxList)
    tim=zeros(size(im9));
    tim(cc.PixelIdxList{i})=1;
    E = bwmorph(tim, 'endpoints');
    [y,x] = find(E);
    pts=[];
    dists=[];
    counter=1;
    pts(counter,:)=[y(1),x(1)];
    tim(pts(counter,1),pts(counter,2))=0;
    while sum(tim(:)>0)
        counter=counter+1;
        [y,x]=find(tim(pts(counter-1,1)+(-1:1),pts(counter-1,2)+(-1:1)));
        dists(counter)=1+(abs(x)==abs(y))*(sqrt(2)-1);
        pts(counter,:)=pts(counter-1,:)+[y,x]-2;
        tim(sub2ind(size(tim),pts(counter,1),pts(counter,2)))=0;
    end %% what does this part do measure the distance?
%     dists
%
    tim=zeros(size(im9));
    rad=box_width/2;
    rad2=box_height/2;
    rad3=box_height*2;
    j=1;
    location=0;
    while(j<(tan_size+1))
        j=j+1;
        location=location+dists(j);
    end
    next_step=location+step_size;
    while(j<=(size(pts,1)-tan_size-1))
        jr=round(j);
        pt=pts(jr,:);
        p1=pts(jr-tan_size,:);
        p2=pts(jr+tan_size,:);
        tan=(p2-p1)/sqrt(sum((p2-p1).^2));

        x=pt(2)+[rad*tan(2)-rad2*tan(1),-rad*tan(2)-rad2*tan(1),-rad*tan(2)+rad2*tan(1),rad*tan(2)+rad2*tan(1)];
        y=pt(1)+[rad*tan(1)+rad2*tan(2),-rad*tan(1)+rad2*tan(2),-rad*tan(1)-rad2*tan(2),rad*tan(1)-rad2*tan(2)];

        x2=pt(2)+[-rad3*tan(1),rad3*tan(1)];
        y2=pt(1)+[rad3*tan(2),-rad3*tan(2)];
        x2=linspace(x2(1),x2(2));
        y2=linspace(y2(1),y2(2));
        y2=y2(and(x2>=1,x2<=size(tim,2)));
        x2=x2(and(x2>=1,x2<=size(tim,2)));
        x2=x2(and(y2>=1,y2<=size(tim,1)));
        y2=y2(and(y2>=1,y2<=size(tim,1)));

        valid=mask(sub2ind(size(im7),round(y2),round(x2)))>0;
        if(sum(valid)<2)
            valid(floor(length(valid)/2)+(1:2))=1;
        end
        x2=x2(valid);
        y2=y2(valid);

        x2=x2([1,end]);
        y2=y2([1,end]);

        dist=sqrt((x2(2)-x2(1))^2+(y2(2)-y2(1))^2);
        n=round((dist-patch_height+step_size2)/(step_size2));
        delta=max(0,dist-n*step_size2-(patch_height-step_size2));

        vec=([x2(2),y2(2)]-[x2(1),y2(1)])/dist;
        x2(1)=x2(1)+vec(1)*delta/2;
        x2(2)=x2(2)-vec(1)*delta/2;
        y2(1)=y2(1)+vec(2)*delta/2;
        y2(2)=y2(2)-vec(2)*delta/2;

        dist=dist-delta;

        tim(sub2ind(size(tim),round(y2),round(x2)))=1;
        vec=([x2(2),y2(2)]-[x2(1),y2(1)])/dist;
        x2(2)=x2(2)-vec(1)*patch_height;
        y2(2)=y2(2)-vec(2)*patch_height;
        tim(sub2ind(size(tim),round(y2),round(x2)))=1;
        x3=linspace(x2(1),x2(2),n);
        y3=linspace(y2(1),y2(2),n);

        for k=1:length(x3)
            x=x3(k)+[rad*tan(2),-rad*tan(2),-rad*tan(2)+2*rad*tan(1),rad*tan(2)+2*rad*tan(1)];
            y=y3(k)+[rad*tan(1),-rad*tan(1),-rad*tan(1)-2*rad*tan(2),rad*tan(1)-2*rad*tan(2)];

            line_color='blue';
            bw=poly2mask(x,y,size(mask,1),size(mask,2));

            if(sum(bw(:).*mask(:))>slide_thresh*sum(bw(:)))
                line(x([1,2]),y([1,2]),'Color',line_color);
                line(x([2,3]),y([2,3]),'Color',line_color);
                line(x([3,4]),y([3,4]),'Color',line_color);
                line(x([4,1]),y([4,1]),'Color',line_color);
                counter2=counter2+1;

                x1=linspace(x(1),x(2),patch_size);
                x2=linspace(x(4),x(3),patch_size);
                y1=linspace(y(1),y(2),patch_size);
                y2=linspace(y(4),y(3),patch_size);

                location2=zeros(patch_size,patch_size,2);
                for m=1:length(x1)
                    x4=round(linspace(x1(m),x2(m),patch_size)*scale);
                    y4=round(linspace(y1(m),y2(m),patch_size)*scale);
                    location2(m,:,1)=x4;
                    location2(m,:,2)=y4;
                end
                locations{counter2}=location2;
            end
        end

        if(j==(size(pts,1)-tan_size-1))
            j=j+1;
            location=location+dists(j);
        end
        while(and(location<next_step,j<=(size(pts,1)-tan_size-1)))
            j=j+1;
            location=location+dists(j);
        end
        next_step=next_step+step_size;
    end % for while in line 136
end % for for in line 106